credoai/modules/model_modules/security.py

```diff
@@ -0,0 +1,145 @@
+import copy
+import warnings
+import numpy as np
+import pandas as pd
+import tensorflow as tf
+from art.attacks.extraction import CopycatCNN
+from art.estimators.classification import KerasClassifier
+from art.estimators.classification.scikitlearn import SklearnClassifier
+from credoai.modules.credo_module import CredoModule
+from credoai.utils.common import NotRunError
+from keras.layers import Dense
+from keras.models import Sequential
+from keras.utils.np_utils import to_categorical
+from sklearn import metrics as sk_metrics
+
+tf.compat.v1.disable_eager_execution()
+import os
+os.environ['TF_CPP_MIN_LOG_LEVEL'] = '3' 
+
+class SecurityModule(CredoModule):
+    """Security module for Credo AI.
+
+    This module takes in binary classification model and data and 
+     provides functionality to perform security assessment
+
+    Parameters
+    ----------
+    model : model
+        A trained ML model
+    x_test : pandas.DataFrame
+        The test features
+    y_test : pandas.Series
+        The test outcome labels
+    """
+
+    def __init__(
+        self, model, x_test, y_test
+    ):
+        self.x_test = x_test.to_numpy()
+        self.y_test = to_categorical(y_test, num_classes=2)
+        self.model = model.model
+        self.victim_model = SklearnClassifier(self.model)
+        np.random.seed(10)
+
+    def run(self):
+        """Runs the assessment process
+
+        Returns
+        -------
+        dict
+            Key: metric name
+            Value: metric value
+        """
+        extraction_performance = self._extraction_attack()
+
+        attack_scores = {
+            **extraction_performance
+        }
+
+        self.results = attack_scores
+
+        return self
+
+    def prepare_results(self):
+        """Prepares results for export to Credo AI's Governance App
+
+        Structures a subset of results for export as a dataframe with appropriate structure
+        for exporting. See credoai.modules.credo_module.
+
+        Returns
+        -------
+        pd.Series
+
+        Raises
+        ------
+        NotRunError
+            If results have not been run, raise
+        """
+        if self.results is not None:
+            return pd.Series(self.results, name='value')
+        else:
+            raise NotRunError("Results not created yet. Call 'run' to create results")
+
+    def _extraction_attack(self):
+        """Model extraction security attack
+
+        In model extraction, the adversary only has access to the prediction API of a target model
+            which she queries to extract information about the model internals and train a substitute model.
+
+        Returns
+        -------
+        dict
+            Key: extraction_attack_score
+            Value: accuracy of the thieved model / accuracy of the victim model, corrected for chance
+        """
+        # use half of the test data for model extraction and half for evaluation
```

Comment:
I'm guessing extraction attack success is a function of the amount of data that is used?

Right now the extraction attack will assume that the attacker has access to half X_test. Why are we making that assumption? Why not use cross-validation?

Reply:
I also noticed this issue. A reasonable process would be one in which an attacker obtains highest accuracy with lowest inference queries through smart model design and querying.
But ART requires the feature vectors to be passed beforehand (I think the attack itself should identify the optimal feature vectors to query for, similar to active learning). All the [three ART extraction methods](https://adversarial-robustness-toolbox.readthedocs.io/en/latest/modules/attacks/extraction.html) require that.

Reply:
Makes sense - that means we need to make the choice for them. We are choosing to use half of X_test. We should list these parameterizations somewhere. I'll make a confluence doc.